### PR TITLE
[WIP] PI-1585 Yak: Upgrade deps, add defer ack/nack

### PR DESCRIPTION
**Pull request by jamescarlson**
# Description

- Upgrade dependencies: Typescript 3, amqplib 0.5.3
- Add `defer` function property to the 2nd handler argument `ackFns` for Receive, Subscribe, and ReceiveTypes. This allows synchronous handler functions which call out to async/promise-returning functions to only Ack the handled message once the asynchronous work completes, accomplished by calling `defer` before returning and then calling `ack` or `nack` in a callback or in a `.then`.

Usage example:
```javascript
bus.Suscribe(type, subscriberName, (message, ackFns) => {
  ackFns.defer();
  someAsyncFunction(message).then(() => ackFns.ack());
});
``` 

## TODO

- [ ] Tests

### CODE REVIEW

File: Bus.d.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export declare class Bus implements IBus {
     private pubChanUp;
     private rpcConsumerUp;
     private static remove$type;
-    SendToErrorQueue(msg: any, err?: string, stack?: string): Promise<boolean>;
+    SendToErrorQueue(msg: any, err?: string, stack?: string): any;
```
Comment (sfrooster) [inline Bus.d.ts:23]: why not something more specific than "any"?

**Review by sfrooster — COMMENTED** [review]
I'd be curious to see "defer" in action.  It also seems like there ought to be some kind of defer fail safe so that deferred acks/nacks which don't get ack'd/nack'd (because defer doesn't force this and maybe a dev forgot or it's just taking way too long) will eventually log/throw/nack.  Thoughts?

**Review by sfrooster — COMMENTED** [review]
I think there is a ton of work - a-lot of updating - that can be done to this aged code.  I've been working through a-lot of that as branch including the "defer" stuff.  I wouldn't want that work to hold you up, but I do think it'd be worth considering updating the code you touch (make var go away, things like that).  Beyond that, my biggest comment is that it seems like "nackIfFirstDeliveryElseSendToErrorQueue" could be rolled back into "nack".  Here's an alternative form (disregard my defer implementation because it allows for a timeout value or a promise-returning function as a parameter).  You decide:

```
let ackdOrNackd = false;
let deferred = false;
let deferTimeout: NodeJS.Timeout;

const ack = () => {
    if (deferred) clearTimeout(deferTimeout);
    channel.ack(msg);
    ackdOrNackd = true;
};

const nack = () => {
    if (deferred) clearTimeout(deferTimeout);
    if (!msg.fields.redelivered) {
        channel.nack(msg);
    }
    else {
        //can only nack once
        this.SendToErrorQueue(_msg, "attempted to nack previously nack'd message");
    }
    ackdOrNackd = true;
};

handler(_msg, {
    ack: ack,
    nack: nack,
    defer: (guard = Bus.defaultDeferredAckTimeout) => {
        if (_.isNumber(guard)) {
            if (!_.isSafeInteger(guard)) {
                // TODO - do what?
            }
            deferTimeout = setTimeout(nack, guard);
        }
        else {
            guard().then(success => success ? ack() : nack());
        }
        deferred = true;
    }
});

if (!ackdOrNackd && !deferred) channel.ack(msg);
```

**Comment by jamescarlson** [comment]
Structuring it that way works for me, I'll change it.

**Pull request merged**